x/programs/rust/wasmlanche_sdk/src/simulator.rs

```diff
@@ -0,0 +1,363 @@
+    pub fn run<T>(&self, simulation: &Simulation) -> Result<T, Box<dyn Error>>
+    where
+        T: serde::de::DeserializeOwned + serde::Serialize,
+    {
+        call_run_stdin(&self.path, simulation)
+    }
+
+    /// Performs a view step against the simulator and returns the result.
+    /// # Errors
+    ///
+    /// Returns an error if the if serialization or simulation fails.
+    pub fn view<T>(&self, data: Step, key: &str) -> Result<T, Box<dyn Error>>
+    where
+        T: serde::de::DeserializeOwned + serde::Serialize,
+    {
+        let simulation = &Simulation {
+            name: "view",
+            description: "single view request",
+            caller_key: key,
+            steps: vec![data],
+        };
+
+        call_run_stdin(&self.path, simulation)
+    }
+
+    /// Performs a single execution step against the simulator and returns the result.
+    /// # Errors
+    ///
+    /// Returns an error if the if serialization or simulation fails.
+    pub fn execute<T>(&self, data: Step, key: &str) -> Result<T, Box<dyn Error>>
+    where
+        T: serde::de::DeserializeOwned + serde::Serialize,
+    {
+        let simulation = &Simulation {
+            name: "execute",
+            description: "single execution request",
+            caller_key: key,
+            steps: vec![data],
+        };
+
+        call_run_stdin(&self.path, simulation)
+    }
+}
+
+fn call_run_stdin<T>(path: &str, simulation: &Simulation) -> Result<T, Box<dyn Error>>
+where
+    T: serde::de::DeserializeOwned + serde::Serialize,
+{
+    let mut child = Command::new(path)
+        .arg("run")
+        .stdin(Stdio::piped())
+        .spawn()?;
+
+    // write json to stdin
+    let input =
+        serde_json::to_string(simulation).map_err(|e| format!("failed to serialize json: {e}"))?;
+    if let Some(ref mut stdin) = child.stdin {
+        stdin
+            .write_all(input.as_bytes())
+            .map_err(|e| format!("failed to write to stdin: {e}"))?;
+    }
+
+    let output = child
+        .wait_with_output()
+        .map_err(|e| format!("failed to wait for command to finish: {e}"))?;
+
+    let resp: T = serde_json::from_str(String::from_utf8(output.stdout)?.as_ref())
+        .map_err(|e| format!("failed to parse output to json: {e}"))?;
+    Ok(resp)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn parse_yaml<'a>(yaml_content: &'a str) -> Result<Simulation<'a>, Box<dyn std::error::Error>> {
+        let sim: Simulation<'a> = serde_yaml::from_str(yaml_content)?;
+        Ok(sim)
+    }
+
+    #[test]
+    fn test_parse_key_yaml() {
+        let yaml_content = "
+name: token program
+description: Deploy and execute token program
+caller_key: alice_key
+steps:
+  - description: create alice key
+    endpoint: key
+    method: create
+    params:
+      - name: key name
+        type: ed25519
+        value: alice_key
+  - description: create bob key
+    endpoint: key
+    method: create
+    params:
+      - name: key name
```

Comment:
nit: both `name` sections are the same for alice and bob. We probably want to have different names? Also I imagine the name shouldn't have spaces in it?

Reply:
spaces are valid in yaml. name in params is only for readability I will add a note on that. this will help a lot with multiple params.